flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java

```diff
@@ -1834,9 +1834,23 @@ ContainerLaunchContext setupApplicationMasterContainer(
         // ------------------ Prepare Application Master Container  ------------------------------
 
         // respect custom JVM options in the YAML file
-        String javaOpts = flinkConfiguration.getString(CoreOptions.FLINK_JVM_OPTIONS);
-        if (flinkConfiguration.getString(CoreOptions.FLINK_JM_JVM_OPTIONS).length() > 0) {
-            javaOpts += " " + flinkConfiguration.getString(CoreOptions.FLINK_JM_JVM_OPTIONS);
+        String javaOpts = "";
+        String defaultJvmOpts = flinkConfiguration.getString(CoreOptions.FLINK_DEFAULT_JVM_OPTIONS);
+        String jvmOpts = flinkConfiguration.getString(CoreOptions.FLINK_JVM_OPTIONS);
+        String defaultJmJvmOpts =
+                flinkConfiguration.getString(CoreOptions.FLINK_DEFAULT_JM_JVM_OPTIONS);
+        String jmJvmOpts = flinkConfiguration.getString(CoreOptions.FLINK_JM_JVM_OPTIONS);
+        if (!defaultJvmOpts.isEmpty()) {
+            javaOpts += " " + defaultJvmOpts;
+        }
+        if (!jvmOpts.isEmpty()) {
+            javaOpts += " " + jvmOpts;
+        }
+        if (!defaultJmJvmOpts.isEmpty()) {
+            javaOpts += " " + defaultJmJvmOpts;
+        }
+        if (!jmJvmOpts.isEmpty()) {
+            javaOpts += " " + jmJvmOpts;
         }
```

Comment:
nit: 

Using the `StringBuilder` to build the javaOpts, and we can define a method to avoid repetitive code here.

```
    private void method(StringBuilder sb, Configuration conf, ConfigOption<String> option) {
        final String javaOpts = conf.get(option);
        if(javaOpts == null || javaOpts.isEmpty()) {
            return;
        }
        sb.append(' ');
        sb.append(javaOpts);
    }
```